impy/common.py

```diff
@@ -81,19 +105,73 @@ def __init__(self, lib, event_kinematics, event_frame, nevent, npart,
         if impy_config['pre_slice']:
             info(10, 'Pre-slice enabled.')
             self.selection = slice(None, self.npart)
+            self.charge = self._charge_init
             self._apply_slicing()
         else: 
             info(10, 'Pre-slice disabled.')
             self.selection = slice(None, None)
+            self.charge = self._charge_init
         
         # The default slice only cuts limits the view to the array to
         # to the current number of entries
         self._is_filtered = False
+        self._views_cache = []
 
         # Apply boosts into frame required by user
         self.kin.apply_boost(self, event_frame, impy_config["user_frame"])
         self.event_frame = impy_config["user_frame"]
 
+    
+    def filter_final(self):
+        self.selection = (self.status == 1)
+        self._cache_current_view()
+        self._apply_slicing()
+        return self
+      
+    def filter_charged(self):
+        self.selection = (self.charge != 0)
+        self._cache_current_view()
+        self._apply_slicing()
+        return self
+    
+    def _cache_current_view(self):
+        """Cache the current view of data"""
+        self._views_cache.append(EventData(
+            self._npart, self.p_ids, self.status, self.charge,
+            self.px, self.py, self.pz, self.en, self.m,
+            self.vx, self.vy, self.vz, self.vt,
+            self.pem_arr, self.vt_arr
+        ))
+    
+    def remove_last_filter(self):
```

Comment:
I do not understand the need to remove filters. Users familiar with numpy know that applying a mask is destructive (`b = a[mask]`). This is fine in numpy because numpy does not override the original array.

I think impy should either follow this approach and use masks to perform irreversible reductions like numpy, but then not do them in place. Or you should use [numpy's masked arrays](https://numpy.org/doc/stable/reference/maskedarray.html) to keep the in-place filtering non-destructive and reversible.

There shouldn't be a need to cache previous filtering results, since the only thing that needs to be modified is the array mask.

Reply:
The reason why the implementation of filtering uses cache of previous filtering results and not mask is because currently filtering of views `px`, `py`,... is made in-place. The alternative is to change `px`, `py`, ... to properties, where mask will be applied to the initial view to Fortran data. Another consideration was that properties are usually slower than direct access to arrays + overhead of creation of new masked view at every invocation. However, it is not critical if every filtered view to `px`,... is accessed not so many times, because we create it anyway at least once. Moreover, at every filtering we apply filters to all views in _apply_slicing() method instead of one property we need.

Considering all above I can change `px`, `py`, ... to properties. Then it will be wise to convert  npart to property:

```python
@property
def npart(self):
     return len(self.p_ids)
```
while keeping initial `npart` as private `_npart`. In this case the properties for momenta will look like :

```python
@property
def px(self):
     return self._pem_arr[0, mask]
``` 

or even

```python
@property
def px(self):
     return self.lib.phep[0, mask]
``` 

where self.lib.phep is the fortran array. The layout of `phep` array is px1 py1, pz1,  ... px2 py2, pz1, so the numpy array will be a view with strides.

About remove filters: if we use `mask` then to undone some filtering we should keep a cache of `mask`s. Or do you mean that we don't need the 'undone'/'clear' filtering functions at all?

Reply:
Thank you for the discussion. I would like to experiment a bit with ideas on filtering in a separate PR. I think for this PR the code is fine as it is.